L1Trigger/TrackFindingTMTT/src/Histos.cc

```diff
@@ -0,0 +1,3970 @@
+           << "WARNING: You should update the constants in class DegradeBend, since some stubs had bend outside the "
+              "expected window range."
+           << endl;
+
+    // Check if GP B approximation cfg params are inconsistent.
+    if (bApproxMistake_)
+      cout << endl << "WARNING: BApprox cfg params are inconsistent - see printout above." << endl;
+  }
+
+  //=== Determine "B" parameter, used in GP firmware to allow for tilted modules.
+
+  void Histos::trackerGeometryAnalysis(const TrackerGeometryInfo trackerGeometryInfo) {
+    // Don't bother producing summary if user didn't request histograms via TFileService in their cfg.
+    if (!this->enabled())
+      return;
+
+    cout << endl << "=========================================================================" << endl;
+    cout << "--- Fit to cfg params for FPGA-friendly approximation to B parameter in GP & KF ---" << endl;
+    cout << "--- (used to allowed for tilted barrel modules)                                 ---" << endl;
+    // Check that info on the correct number of modules has been stored
+    if (trackerGeometryInfo.moduleZoR().size() !=
+        trackerGeometryInfo.barrelNTiltedModules() * trackerGeometryInfo.barrelNLayersWithTiltedModules() * 2) {
+      cout << "WARNING : Expected "
+           << trackerGeometryInfo.barrelNTiltedModules() * trackerGeometryInfo.barrelNLayersWithTiltedModules() * 2
+           << " modules, but only recorded info on " << trackerGeometryInfo.moduleZoR().size() << endl;
+    }
+
+    TFileDirectory inputDir = fs_->mkdir("InputData");
+    graphBVsZoverR_ = inputDir.make<TGraph>(
+        trackerGeometryInfo.moduleZoR().size(), &trackerGeometryInfo.moduleZoR()[0], &trackerGeometryInfo.moduleB()[0]);
+    graphBVsZoverR_->SetNameTitle("B vs module Z/R", "; Module Z/R; B");
+    graphBVsZoverR_->Fit("pol1", "q");
+    TF1* fittedFunction = graphBVsZoverR_->GetFunction("pol1");
+    double gradient = fittedFunction->GetParameter(1);
+    double intercept = fittedFunction->GetParameter(0);
+    cout << "         BApprox_gradient (fitted)  = " << gradient << endl;
+    cout << "         BApprox_intercept (fitted) = " << intercept << endl;
+    // Check fitted params consistent with those assumed in cfg file.
+    if (settings_->useApproxB()) {
+      double gradientDiff = fabs(gradient - settings_->bApprox_gradient());
+      double interceptDiff = fabs(intercept - settings_->bApprox_intercept());
+      if (gradientDiff > 0.001 || interceptDiff > 0.001) {  // Uncertainty independent of number of events
+        cout << endl << "WARNING: fitted parameters inconsistent with those specified in cfg file:" << endl;
+        cout << "         BApprox_gradient  (cfg) = " << settings_->bApprox_gradient() << endl;
+        cout << "         BApprox_intercept (cfg) = " << settings_->bApprox_intercept() << endl;
+        bApproxMistake_ = true;  // Note that problem has occurred.
+      }
+    }
+  }
+
+}  // namespace TMTT
```

Comment:
They are purely used for histogramming to optimize & understand the L1 tracking, and are split up into functions within the file to separate histograms used to analyze different parts of the chain.

Reply:
Only a dummy version of class Histos, with no implementation, will be included in the updated PR. (This will allow us to replace it will the full implementation for single-threaded L1 track development work, but will avoid the problem of histograms preventing multithreaded EDProducers for CMSSW official code)